File: examples/shapes/implementations/ruby/lib/shapes.rb
# frozen_string_literal: true

require_relative "shapes/version"

module Shapes
  class Shape
    def area
      raise NotImplementedError, "Subclass must implement abstract method"
    end

    def perimeter
      raise NotImplementedError, "Subclass must implement abstract method"
    end
  end

  class Polygon < Shape
    def edge_count
      raise NotImplementedError, "Subclass must implement abstract method"
    end

    def vertex_count
      raise NotImplementedError, "Subclass must implement abstract method"
    end
  end

  class Circle < Shape
    attr_reader :radius

    def initialize(radius)
      @radius = radius
    end

    def area
      Math::PI * @radius * @radius
    end

    def perimeter
      2 * Math::PI * @radius
    end

    def diameter
      2 * @radius
    end
  end

  class Rectangle < Polygon
    attr_reader :width, :height

    def initialize(width, height)
      @width = width
      @height = height
    end

    def area
      @width * @height
    end

    def perimeter
      2 * (@width + @height)
    end

    def edge_count
      4
    end

    def vertex_count
      4
    end

    def square?
      @width == @height
    end
  end

  class Triangle < Polygon
    attr_reader :side_a, :side_b, :side_c

    def initialize(side_a, side_b, side_c)
      @side_a = side_a
      @side_b = side_b
      @side_c = side_c
    end

    def area
      s = (@side_a + @side_b + @side_c) / 2.0
      Math.sqrt(s * (s - @side_a) * (s - @side_b) * (s - @side_c))
    end

    def perimeter
      @side_a + @side_b + @side_c
    end

    def edge_count
      3
    end

    def vertex_count
      3
    end
  end
end

File: examples/shapes/implementations/swift/Sources/Shapes/Shapes.swift
// The Swift Programming Language
// https://docs.swift.org/swift-book

import Foundation

protocol Shape {
    func area() -> Double
    func perimeter() -> Double
}

protocol Polygon: Shape {
    func edgeCount() -> Int
    func vertexCount() -> Int
}

public class Circle: Shape {
    public let radius: Double

    public init(radius: Double) {
        self.radius = radius
    }

    public func area() -> Double {
        return Double.pi * radius * radius
    }

    public func perimeter() -> Double {
        return 2 * Double.pi * radius
    }

    public func diameter() -> Double {
        return 2 * radius
    }
}

public class Rectangle: Polygon {
    public let width: Double
    public let height: Double

    public init(width: Double, height: Double) {
        self.width = width
        self.height = height
    }

    public func area() -> Double {
        return width * height
    }

    public func perimeter() -> Double {
        return 2 * (width + height)
    }

    public func edgeCount() -> Int {
        return 4
    }

    public func vertexCount() -> Int {
        return 4
    }

    public func isSquare() -> Bool {
        return width == height
    }
}

public class Triangle: Polygon {
    public let sideA: Double
    public let sideB: Double
    public let sideC: Double

    public init(sideA: Double, sideB: Double, sideC: Double) {
        self.sideA = sideA
        self.sideB = sideB
        self.sideC = sideC
    }

    public func area() -> Double {
        let s = (sideA + sideB + sideC) / 2
        return sqrt(s * (s - sideA) * (s - sideB) * (s - sideC))
    }

    public func perimeter() -> Double {
        return sideA + sideB + sideC
    }

    public func edgeCount() -> Int {
        return 3
    }

    public func vertexCount() -> Int {
        return 3
    }
}
